### PR TITLE
Issue #1811: Add static methods to construct Future class using Calla…

### PR DESCRIPTION
**Pull request by v1ctor**
…ble, Supplier, Runnable
#1811

### CODE REVIEW

File: javaslang/src/main/java/javaslang/concurrent/Future.java [hunk 1]
```diff
@@ -323,6 +324,79 @@
      * Starts an asynchronous computation, backed by the {@link #DEFAULT_EXECUTOR_SERVICE}.
      *
      * @param computation A computation.
+     * @param <T>      Type of the computation result.
+     * @return A new Future instance
+     * @throws NullPointerException if computation is null.
+     */
+    static <T> Future<T> ofSupplier(Supplier<? extends T> computation) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/concurrent/Future.java]: Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

File: javaslang/src/main/java/javaslang/concurrent/Future.java [hunk 2]
```diff
@@ -323,6 +324,79 @@
...
+    /**
+     * Starts an asynchronous computation, backed by the given {@link ExecutorService}.
+     *
+     * @param executorService An executor service.
+     * @param computation A computation.
+     * @param <T>      Type of the computation result.
+     * @return A new Future instance
+     * @throws NullPointerException if one of executorService or computation is null.
+     */
+    static <T> Future<T> ofSupplier(ExecutorService executorService, Supplier<? extends T> computation) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/concurrent/Future.java]: Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

File: javaslang/src/main/java/javaslang/concurrent/Future.java [hunk 3]
```diff
@@ -323,6 +324,79 @@
...
+
+    /**
+     * Starts an asynchronous computation, backed by the {@link #DEFAULT_EXECUTOR_SERVICE}.
+     *
+     * @param computation A computation
+     * @param <T>      Type of the computation result.
+     * @return A new Future instance
+     * @throws NullPointerException if computation is null.
+     */
+    static <T> Future<T> ofCallable(Callable<? extends T> computation) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/concurrent/Future.java]: Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

File: javaslang/src/main/java/javaslang/concurrent/Future.java [hunk 4]
```diff
@@ -323,6 +324,79 @@
...
+    /**
+     * Starts an asynchronous computation, backed by the given {@link ExecutorService}.
+     *
+     * @param executorService An executor service.
+     * @param computation A computation.
+     * @param <T>      Type of the computation result.
+     * @return A new Future instance
+     * @throws NullPointerException if one of executorService or computation is null.
+     */
+    static <T> Future<T> ofCallable(ExecutorService executorService, Callable<? extends T> computation) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/concurrent/Future.java]: Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

**Comment by v1ctor** [comment]
Hi Daniel, thanks for the feedback! First I added null checks to the code, but then I realized that as soon as these methods are wrappers around existing functions these checks will be executed inside. So I removed them. Do you want me to add them explicitly?

**Comment by danieldietrich** [comment]
Hi

> Do you want me to add them explicitly?

yes, please. Only for the mentioned method references. The `paramName::methodName` method references throw before passed as new parameter to an existing function.

The executorService does not need to be checked again.

---

**Example:**

```java
import java.util.Objects;
import java.util.function.Supplier;

public class Test {

    public static void main(String[] args) {
        testSupplier(null);
    }

    static <T> void testSupplier(Supplier<T> s) {
        testCheckedSupplier(s::get); // <------ throws immediately if s is null
    }

    static <T> void testCheckedSupplier(Supplier<T> s) {
        Objects.requireNonNull(s, "s is null");
        System.out.println("ok");
    }

}
```

**Output:**

```java
Exception in thread "main" java.lang.NullPointerException
	at javaslang.Test.testSupplier(Test.java:18)
	at javaslang.Test.main(Test.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

**Comment by v1ctor** [comment]
@danieldietrich Yes, you're totally right. I've missed this case. Thanks!

**Comment by danieldietrich** [comment]
Thank you!

**Comment by v1ctor** [comment]
Thank you :)

**Comment by danieldietrich** [comment]
Fixes #1811

**Pull request merged**